pkg/controller/deployment/deployment_controller.go

```diff
@@ -709,24 +707,43 @@ func (dc *DeploymentController) scaleUpNewRCForRecreate(newRC *api.ReplicationCo
 }
 
 func (dc *DeploymentController) updateDeploymentStatus(allRCs []*api.ReplicationController, newRC *api.ReplicationController, deployment extensions.Deployment) error {
-	totalReplicas := deploymentutil.GetReplicaCountForRCs(allRCs)
-	updatedReplicas := deploymentutil.GetReplicaCountForRCs([]*api.ReplicationController{newRC})
-	availablePods, err := deploymentutil.GetAvailablePodsForRCs(dc.client, allRCs, deployment.Spec.Strategy.RollingUpdate.MinReadySeconds)
+	totalReplicas, updatedReplicas, availableReplicas, unavailableReplicas, err := dc.calculateStatus(allRCs, newRC, deployment)
 	if err != nil {
-		return fmt.Errorf("failed to count ready pods: %v", err)
+		return err
 	}
 	newDeployment := deployment
 	// TODO: Reconcile this with API definition. API definition talks about ready pods, while this just computes created pods.
 	newDeployment.Status = extensions.DeploymentStatus{
 		Replicas:            totalReplicas,
 		UpdatedReplicas:     updatedReplicas,
-		AvailableReplicas:   availablePods,
-		UnavailableReplicas: totalReplicas - availablePods,
+		AvailableReplicas:   availableReplicas,
+		UnavailableReplicas: unavailableReplicas,
 	}
 	_, err = dc.expClient.Deployments(deployment.ObjectMeta.Namespace).UpdateStatus(&newDeployment)
 	return err
 }
 
+func (dc *DeploymentController) calculateStatus(allRCs []*api.ReplicationController, newRC *api.ReplicationController, deployment extensions.Deployment) (totalReplicas, updatedReplicas, availableReplicas, unavailableReplicas int, err error) {
+	totalReplicas = deploymentutil.GetReplicaCountForRCs(allRCs)
+	updatedReplicas = deploymentutil.GetReplicaCountForRCs([]*api.ReplicationController{newRC})
+	switch deployment.Spec.Strategy.Type {
+	case extensions.RecreateDeploymentStrategyType:
+		availableReplicas, err = deploymentutil.GetAvailablePodsForRCs(dc.client, allRCs, 0)
+		if err != nil {
+			err = fmt.Errorf("failed to count available pods: %v", err)
+			return
+		}
+	case extensions.RollingUpdateDeploymentStrategyType:
+		availableReplicas, err = deploymentutil.GetAvailablePodsForRCs(dc.client, allRCs, deployment.Spec.Strategy.RollingUpdate.MinReadySeconds)
```

Comment:
If the two cases are identical except for the last parameter, I suggest just setting a variable in the switch and moving the call to GetAvailablePodsForRCs below.